agent/taskresource/ssmsecret/ssmsecret.go

```diff
@@ -0,0 +1,476 @@
+		wg.Add(1)
+		go secret.retrieveSSMSecretValuesByRegion(region, secrets, iamCredentials, &wg, errorEvents)
+	}
+
+	wg.Wait()
+	//get the first error returned and set as terminal reason
+	select {
+	case err := <-errorEvents:
+		secret.setTerminalReason(err.Error())
+		return err
+	default:
+		break
+	}
+	return nil
+}
+
+func (secret *SSMSecretResource) getGoRoutineTotalNum() int {
+	total := 0
+	for _, secrets := range secret.requiredSecrets {
+		total += len(secrets)/10 + 1
+	}
+	return total
+}
+
+func (secret *SSMSecretResource) retrieveSSMSecretValuesByRegion(region string, secrets []apicontainer.Secret, iamCredentials credentials.IAMRoleCredentials, wg *sync.WaitGroup, errorEvents chan error) {
+	seelog.Infof("ssm secret resource: retrieving secrets for region %s in task: [%s]", region, secret.taskARN)
+	defer wg.Done()
+
+	var wgPerRegion sync.WaitGroup
+	var secretNames []string
+
+	for _, s := range secrets {
+		secretName := secret.extractNameFromValueFrom(s)
+		secretKey := secretName + "_" + region
+		if _, ok := secret.GetSSMSecretValue(secretKey); ok {
+			continue
+		}
+		secretNames = append(secretNames, secretName)
+		if len(secretNames) == 10 {
+			secretNamesTmp := make([]string, 10)
+			copy(secretNames, secretNamesTmp)
+			wgPerRegion.Add(1)
+			go secret.retrieveSSMSecretValues(region, secretNamesTmp, iamCredentials, &wgPerRegion, errorEvents)
+			secretNames = []string{}
+		}
+	}
+
+	if len(secretNames) > 0 {
+		wgPerRegion.Add(1)
+		go secret.retrieveSSMSecretValues(region, secretNames, iamCredentials, &wgPerRegion, errorEvents)
+	}
+
+	wgPerRegion.Wait()
+	return
+}
+
+func (secret *SSMSecretResource) retrieveSSMSecretValues(region string, names []string, iamCredentials credentials.IAMRoleCredentials, wg *sync.WaitGroup, errorEvents chan error) {
+	defer wg.Done()
+
+	ssmClient := secret.ssmClientCreator.NewSSMClient(region, iamCredentials)
+	seelog.Infof("ssm secret resource: retrieving resource for secrets %v in region [%s] in task: [%s]", names, region, secret.taskARN)
+	secValueMap, err := ssm.GetSecretsFromSSM(names, ssmClient)
+	if err != nil {
+		errorEvents <- err
+		return
+	}
+
+	secret.lock.Lock()
+	defer secret.lock.Unlock()
+
+	// put secret value in secretData
+	for secretName, secretValue := range secValueMap {
+		secretKey := secretName + "_" + region
+		secret.secretData[secretKey] = secretValue
+	}
+
+	return
+}
+
+func (secret *SSMSecretResource) extractNameFromValueFrom(secretData apicontainer.Secret) string {
+	valueFrom := secretData.ValueFrom
+	match, _ := regexp.MatchString(ParameterARNRegex, valueFrom)
+
+	if match {
+		index := strings.Index(valueFrom, ParameterARNPrefix)
+		return string(valueFrom[index+len(ParameterARNPrefix):])
+	} else {
+		return valueFrom
+	}
+
+}
+
+// GetRequiredSecrets returns the requiredSecrets field of ssmsecret task resource
+func (secret *SSMSecretResource) GetRequiredSecrets() map[string][]apicontainer.Secret {
```

Comment:
This and `GetExecutionCredentialsID` need not be a public method?

Reply:
I use public method here, so it can also used outside package if we need. It is one of those get/set functions to retrieve/set attributes outside packages. Even so far, it is only used within the package.

Reply:
I believe the resource is accessed only using the interface methods outside the package. If it is not used outside the package, let's make them private.

Reply:
I see ur point, updated to private